v2/codegen/infragen/configgen/configgen.go

```diff
@@ -384,7 +394,24 @@ func (cb *configUnmarshalersBuilder) typeUnmarshalerFunc(typ schema.Type) (f *St
 // typeUnmarshalerName returns a generated name for the unmarshaler function for the given type and the type that
 // the decl is.
 func (cb *configUnmarshalersBuilder) typeUnmarshalerName(decl *schema.TypeDecl) (reader string, rtnTyp *Statement) {
-	return fmt.Sprintf("encoreInternalTypeConfigUnmarshaler_%s", decl.Name), Id(decl.Name)
+	rtnTyp = genutil.Q(decl.Info)
+
+	// If we've already allocated a name for this type, reuse it.
+	key := decl.Info.QualifiedName()
+	if allocated, ok := cb.allocatedNames[key]; ok {
+		return allocated, rtnTyp
+	}
+
+	// Otherwise allocate a new name.
+	suffix := ""
+	if n := cb.seenNames[decl.Name]; n > 0 {
+		suffix = fmt.Sprintf("%d", n+1)
+	}
+	name := fmt.Sprintf("encoreInternalTypeConfigUnmarshaler_%s%s", decl.Name, suffix)
+	cb.seenNames[decl.Name]++
+	cb.allocatedNames[key] = name
+
+	return name, rtnTyp
```

Comment:
Nit; While this works, I wonder if it'd be better to also key off the package name too? - although we'd still need the numerical version fallback incase we saw `pkg_bar.Foo` and `pkg.bar_foo` and `pkg_bar_foo` at the same time.